torch/lib/c10d/ProcessGroupNCCL.cpp

```diff
@@ -916,6 +918,13 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupNCCL::WorkNCCL::
       outputs_->size() == 1,
       "WorkNCCL's getFuture API is only supported for single-process single-device mode.");
   auto deviceIndex = (*outputs_)[0].device().index();
+  // Do not free the resluting tensors before their usages on
+  // futureNCCLCallbackStream_ finish.
+  for (const auto& tensor : *outputs_) {
+    c10::cuda::CUDACachingAllocator::recordStream(
+        tensor.storage().data_ptr(), *futureNCCLCallbackStreams_[deviceIndex]);
+  }
```

Comment:
Don't think we need recordStream here since we aren't actually running on the futureNCCLCallbackStream here.